File: solutions/node/src/adapters/persistence/InMemoryChargingSessionRepository.ts
import { ChargingSession, SessionStatus } from '../../charging/domain/ChargingSession.ts';
import type { ChargingSessionRepository } from '../../charging/application/ports.ts';

// Simplest adapter that satisfies the port. Day 6 replaces it with SQLite/Postgres
// (see docker/init.sql for the table shape) without touching domain or application code.
export class InMemoryChargingSessionRepository implements ChargingSessionRepository {
  private readonly bySessionId = new Map<string, ChargingSession>();

  async findById(sessionId: string): Promise<ChargingSession | undefined> {
    return this.bySessionId.get(sessionId);
  }

  async findActiveByConnector(connectorId: string): Promise<ChargingSession | undefined> {
    for (const session of this.bySessionId.values()) {
      if (session.connectorId === connectorId && session.status === SessionStatus.Charging) return session;
    }
    return undefined;
  }

  async save(session: ChargingSession): Promise<void> {
    if (!session.sessionId) return; // a rejected/never-started session has nothing to persist
    this.bySessionId.set(session.sessionId, session);
  }

  all(): ChargingSession[] {
    return [...this.bySessionId.values()];
  }
}

File: starter/python/src/adapters/persistence/in_memory_charging_session_repository.py
"""Simplest adapter that satisfies the port. Day 6 replaces it with SQLite/Postgres
(see docker/init.sql for the table shape) without touching domain or application code."""
from charging.domain.charging_session import ChargingSession, SessionStatus


class InMemoryChargingSessionRepository:
    def __init__(self) -> None:
        self._by_session_id: dict[str, ChargingSession] = {}

    def find_by_id(self, session_id: str) -> ChargingSession | None:
        return self._by_session_id.get(session_id)

    def find_active_by_connector(self, connector_id: str) -> ChargingSession | None:
        for session in self._by_session_id.values():
            if session.connector_id == connector_id and session.status is SessionStatus.CHARGING:
                return session
        return None

    def save(self, session: ChargingSession) -> None:
        if session.session_id is None:
            return  # a rejected/never-started session has nothing to persist
        self._by_session_id[session.session_id] = session

    def all(self) -> list[ChargingSession]:
        return list(self._by_session_id.values())
